jitstats.lua

```diff
@@ -0,0 +1,462 @@
+        stats.loop_blacklisted = stats.loop_blacklisted + 1
+    end
+
+    -- We sometimes don't can't find bc
+    bc = funcbc(func, pc)
+    local opcode = bc and band(bc, 0xff)
+    
+    if code == TRERR_LLEAVE then
+        debugprint("LLEAVE: %d/%s", opcode, bcnames[opcode])
+    elseif code == TRERR_LUNROLL then
+        
+    elseif code == TRERR_NYIBC then
+        bcname = bcnames[errinfo]
+        stats.nyi_bc[bcname] = (stats.nyi_bc[bcname] or 0) + 1
+    end
+
+    -- Use internal enum name of the trace abort error if possible
+    local tracerror = traceerr_fromid[code] or vmdef.traceerr[code]
+    stats.abort_counts[tracerror] = (stats.abort_counts[tracerror] or 0) + 1
+end
+
+function vmevent.flush()
+    debugprint("Full trace flush")
+    stats.traceflush = stats.traceflush + 1
+end
+
+function vmevent.texit(id, exitno)
+    debugprint("EXIT(%d): exit = %d", id, exitno)
+    stats.exits = stats.exits + 1
+end
+
+local function vmeventcb(what, ...)
+    local cb = vmevent[what]
+    if cb then 
+        cb(...)
+    end
+end
+
+function jitstats.start()
+    if jitstats.isrunning then
+        return false
+    end
+
+    jit.attach(vmeventcb, "trace")
+    jit.attach(vmevent.texit, "texit")
+    jitstats.isrunning = true
+    return true
+end
+
+function jitstats.stop()
+    if not jitstats.isrunning then
+        return false
+    end
+
+    jit.attach(vmeventcb)
+    jit.attach(vmevent.texit)
+    jitstats.isrunning = false
+    return true
+end
+
+local last_starts, last_exits
+
+function jitstats.reset(statstbl)
+    statstbl = statstbl or stats
+    statstbl.starts = 0
+    statstbl.aborts = 0
+    statstbl.abort_counts = {}
+    statstbl.completed = 0
+    statstbl.linktypes = {}
+    statstbl.stitch_starts = 0
+    statstbl.stitch_aborts = 0
+    statstbl.stitch_completed = 0
+    statstbl.side_starts = 0
+    statstbl.side_aborts = 0  
+    statstbl.side_completed = 0
+    statstbl.nyi_bc = {}
+    statstbl.exits = 0
+    statstbl.traceflush = 0
+    statstbl.func_blacklisted = 0
+    statstbl.loop_blacklisted = 0
+end
+
+local function table_copy(src, dst)
+    dst = dst or {}
+    for k,v in pairs(src) do
+        if type(v) ~= "table" then
+            dst[k] = v
+        else
+            dst[k] = table_copy(v, dst[k])
+        end
+    end
+    return dst
+end
+
+local function initorclear(tab, key)
+    tab[key] = tab[key] or {}
+    table.clear(tab[key])
+end
+
+--Optionally pass in a table to copy to snapshot to instead of creating a new table
+function jitstats.getsnapshot(dst)
+    if dst then
+        initorclear(dst, "nyi_bc")
+        initorclear(dst, "abort_counts")
+        initorclear(dst, "linktypes")
+    end
+
+    return table_copy(stats, dst)
+end
+
+local function diffstatbl(old, new, result, key)
+    assert(old[key], "missing old sub table")
+    assert(new[key], "missing new sub table")
+    assert(result[key], "missing old sub table")
```

Comment:
these asserts can't fire in the current use.  while programming defensively is usually good, on short functions that are used on only one way it can obscure the intention.
two alternatives:
- just get rid of it.  reduce the scope of the function so there's no temptation to use somewhere else where the contract isn't followed.
- change the interface, removing the `key` parmenter.  something like this (untested!):
```
do
	local function diffstatbl(old, new)
		if not new then return old end -- or table_copy(old)  ?
		if type(new) == 'number' then
			return v - (old or 0)
		end
		local r = {}
		for k, v in pairs(new) do
			r[k] = v - (old[k] or 0)
		end
		return r
	end
	
	function jitstats.diffsnapshots(snap1, snap2)
		return {
			abort_counts = diffstatbl(snap1.abort_counts, snap2.abort_counts)
			nyi_bc = diffstatbl(snap1.nyi_bc, snap2.nyi_bc)
			linktypes = diffstatbl(snap1.linktypes, snap2.linktypes)
		}
	end
end
```

Reply:
They did fire for me before because I didn't assign one of the tables in another function. I also did have it without the key parameter before but it felt error prone and verbose to type out the field 3 times when adding a new sub table.

Reply:
yes, it's mostly a style preference, and what anybody considers more readable vs. safe.  in my opinion, if a function is just a helper, used in a single place, very close to where it's defined, and with the scope clearly limited, then safety can be relaxed a bit.

if it were a generic utility, or (much more!) an exported function, then yes, do it very defensively!

but even in that case, if the result of the user not following the contract (i.e. giving a string instead of a table) is a simple and obvious error (i.e. `"attempt to index local 's' (a string value)"`), then i wouldn't be worried.